### PR TITLE
Put endpoint end-to-end JSON tests

### PR DESCRIPTION
**Pull request by guscarreon**
A Prebid-Server-like JSON test framework that allows for robust end-to-end testing.

### CODE REVIEW

**Review by bsardo — COMMENTED** [review]
This is looking good. Just some minor comments around formatting and such.

File: endpoints/put_test.go [hunk 1]
```diff
@@ -16,14 +17,245 @@ import (
...
+				"sample-requests/putEndpointTests/invalid-value/value_missing.json",
+				"sample-requests/putEndpointTests/invalid-value/value_greater_than_max.json",
+			},
+		},
+		{
+			desc:        "Valid when storing under custom keys is allowed, expect 200 responses",
+			expectError: false,
+			tests: []string{
+				"sample-requests/putEndpointTests/custom_keys/allowed/exemplary/key_field_included.json",
+				"sample-requests/putEndpointTests/custom_keys/allowed/exemplary/key_field_missing.json",
```
[the remark this answers is not in the document]
Comment (guscarreon) [reply endpoints/put_test.go:80]: Good catch, removed

File: endpoints/sample-requests/putEndpointTests/custom_keys/not_allowed/exemplary/key_field_missing.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
```
[the remark this answers is not in the document]
Comment (guscarreon) [reply endpoints/sample-requests/putEndpointTests/custom_keys/not_allowed/exemplary/key_field_missing.json:1]: Good catch. Removed unnecessary test

File: endpoints/sample-requests/putEndpointTests/valid-whole/valid_type_xml.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+       {
+         "type": "xml",
+         "ttlseconds": 60,
+         "value": "<?xml version=\"1.0\">__video_info__<\\/VAST>\r\n"
+       }
+    ]
+  },
+  "expectedMetrics": {
+	"totalRequests": 1,
+	"requestDur": 1.00
```
[the remark this answers is not in the document]
Comment (guscarreon) [reply endpoints/sample-requests/putEndpointTests/valid-whole/valid_type_xml.json:14]: Yep, they were all over the place. Corrected.

File: endpoints/put_test.go [hunk 4]
```diff
@@ -16,14 +17,245 @@ import (
...
+	RequestErrs    int64   `json:"requestErrs"`
+	RequestDur     float64 `json:"requestDur"`
+}
+
+type testConfig struct {
+	AllowSettingKeys  bool `json:"custom_keys"`
+	ValueMaxSizeBytes int  `json:"max_size_bytes"`
+	MaxNumValues      int  `json:"max_num_values"`
+	MaxTTLSeconds     int  `json:"max_ttl_seconds"`
+}
```
[the remark this answers is not in the document]
Comment (guscarreon) [reply endpoints/put_test.go:225]: Good catch. Renamed variables and entries in the JSON files to things that made more sense and always in camel case with the exception of `"allow_setting_keys"`, `"max_size_bytes"`, `"max_num_values"`, and `"max_ttl_seconds"`. These last 4 values make more sense to be to me to be identical than those of the YAML configuration file fields. What do you think?

File: endpoints/put_test.go [hunk 5]
```diff
@@ -16,14 +17,245 @@ import (
...
+	testGroups := []struct {
+		desc        string
+		expectError bool
+		tests       []string
+	}{
+		{
+			desc:        "Valid put requests. Expect 200 response",
+			expectError: false,
+			tests: []string{
+				"sample-requests/putEndpointTests/valid-whole/single_element_to_store.json",
```
[the remark this answers is not in the document]
Comment (guscarreon) [reply endpoints/put_test.go:43]: Good point. The casing was all over the place. Changed to `"sample-requests/put-endpoint/valid-whole/single-element-to-store.json"`. Do you agree?

**Review by bsardo — DISMISSED** [review]
LGTM

**Review by AlexBVolcy — COMMENTED** [review]
This looks great! I had a minor comment about test coverage in `put.go`. Is there a way to cover the error cases on lines 134, 166, and 191? If there is could, could you add a couple test cases for them? If not, no worries, and I'll approve.

File: endpoints/put.go [hunk 6]
```diff
@@ -121,6 +121,11 @@ func parsePutObject(p putObject) (string, error) {
 		return "", utils.NewPBCError(utils.NEGATIVE_TTL, fmt.Sprintf("ttlseconds must not be negative %d.", p.TTLSeconds))
 	}
 
+	// Make sure data type is specified
+	//if len(p.Type) == 0 {
+	//	return "", utils.NewPBCError(utils.MISSING_TYPE)
+	//}
+
```
Comment (AlexBVolcy) [inline endpoints/put.go:128]: Could this commented out code be deleted?

File: endpoints/put_test.go [hunk 7]
```diff
@@ -677,50 +947,58 @@ func TestEmptyPutRequests(t *testing.T) {
...
 		rr := httptest.NewRecorder()
 
 		// Create request everytime
-		request, err := http.NewRequest("POST", "/cache", strings.NewReader(test.reqBody))
+		request, err := http.NewRequest("POST", "/cache", strings.NewReader(tc.reqBody))
 		assert.NoError(t, err, "[%d] Failed to create a POST request: %v", i, err)
 
 		// Run
 		router.ServeHTTP(rr, request)
 		//assert.Equal(t, http.StatusOK, rr.Code, "[%d] ServeHTTP(rr, request) failed = %v \n", i, rr.Result())
```
Comment (AlexBVolcy) [inline endpoints/put_test.go:985]: Can this commented out code be deleted? I know this wasn't an addition to this PR, but the clean up could be nice.

**Review by AlexBVolcy — COMMENTED** [review]
Few minor comments!

File: endpoints/put.go (hunk 6 above)
Comment (guscarreon) [reply endpoints/put.go:128]: Removed. Also corrected the `%v` string formatter in line 141 below:
```
140       } else {
141 -         return "", utils.NewPBCError(utils.UNSUPPORTED_DATA_TO_STORE, fmt.Sprintf("Type must be one of [\"json\", \"xml\"]. Found %v", p.Type))
    -         return "", utils.NewPBCError(utils.UNSUPPORTED_DATA_TO_STORE, fmt.Sprintf("Type must be one of [\"json\", \"xml\"]. Found '%s'", p.Type))
142       }
143  
144       return toCache, nil
145   }
```

File: endpoints/put_test.go (hunk 7 above)
Comment (guscarreon) [reply endpoints/put_test.go:985]: Forgot to clean up a little bit. Removed.

**Review by bsardo — APPROVED** [review]
Nice Work. LGTM

**Review by AlexBVolcy — APPROVED** [review]

**Pull request merged**